### PR TITLE
Transaction gossiping

### LINKED ISSUE #3: Transaction gossiping
Transactions will be propagated in the P2P network using the gossiping mechanism.

https://github.com/libp2p/go-libp2p-pubsub provides `Gossipsub` implementation, which looks very interesting.

### PR DESCRIPTION
**Pull request by tzdybal**
Transaction gossiping, using
* lazyledger-core/tendermint mempool
* libp2p GossipSub on P2P layer

Resolves #3.
Depends on #17 and #30.

### CODE REVIEW

File: docs/lazy-adr/adr-002-mempool.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+For now, mempool implementation from lazyledger-core/Tendermint will be used.
+
+## Pros
+* good integration with other re-used code (see ADR-001)
+* well tested
+* glue code is not required
+* it will be updated in case of ABCI++ adoption
+* mempool doesn't depend on P2P layer, so it's easy to replace it with libp2p
+
```
Comment (liamsi) [inline docs/lazy-adr/adr-002-mempool.md:11]: another pro which is implicitly already mentioned in "good integration with other re-used code" is that the mempool does not require any knowledge about the internal structure of the Txs and is already "abci-ready"

File: p2p/client_test.go [hunk 2]
```diff
@@ -112,6 +113,35 @@ func TestDiscovery(t *testing.T) {
...
+
+	time.Sleep(5 * time.Second)
+
+	// gossip from client 4
+	err := clients[4].GossipTx(ctx, []byte("sample tx"))
+	assert.NoError(err)
+
+	nCtx, nCancel := context.WithTimeout(ctx, 1*time.Second)
+	defer nCancel()
+	msg, err := clients[3].txSub.Next(nCtx)
```
Comment (liamsi) [inline p2p/client_test.go:140]: Is this where this times out?

File: p2p/client_test.go [hunk 3]
```diff
@@ -112,6 +113,35 @@ func TestDiscovery(t *testing.T) {
...
+		4: []int{2},
+	}, logger)
+
+	// wait for clients to finish refreshing routing tables
+	clients.WaitForDHT()
+
+	time.Sleep(5 * time.Second)
+
+	// gossip from client 4
+	err := clients[4].GossipTx(ctx, []byte("sample tx"))
```
Comment (liamsi) [inline p2p/client_test.go:135]: not sure what is going on but `pub.ready == nil`. Does that mean this topic gets never bootstrapped? https://github.com/libp2p/go-libp2p-pubsub/blob/5457a2845b06d859df6d3683867808e46a15c500/topic.go#L240-L242

Comment (liamsi) [inline p2p/client_test.go:135]: So the problem seems to be that the message we are trying to publish does not get validate properly: https://github.com/libp2p/go-libp2p-pubsub/blob/5457a2845b06d859df6d3683867808e46a15c500/validation.go#L242-L247

File: p2p/client_test.go [hunk 4]
```diff
@@ -119,25 +114,25 @@ func TestGossiping(t *testing.T) {
...
 
-	time.Sleep(5 * time.Second)
+	time.Sleep(10*time.Second)
 
 	// gossip from client 4
 	err := clients[4].GossipTx(ctx, []byte("sample tx"))
 	assert.NoError(err)
 
-	nCtx, nCancel := context.WithTimeout(ctx, 1*time.Second)
+	nCtx, nCancel := context.WithTimeout(ctx, 300*time.Second)
```
Comment (liamsi) [inline p2p/client_test.go:133]: Wowser. Does that mean, the messages gets now written to the channel but the client isn't able to retrieve it now?

File: p2p/client.go [hunk 5]
```diff
@@ -13,15 +13,22 @@ import (
...
 
 	"github.com/lazyledger/optimint/config"
 	"github.com/lazyledger/optimint/log"
 )
 
-// reAdvertisePeriod defines a period after which P2P client re-attempt advertising namespace in DHT.
-const reAdvertisePeriod = 1 * time.Hour
+const (
+	// reAdvertisePeriod defines a period after which P2P client re-attempt advertising namespace in DHT.
+	reAdvertisePeriod = 1 * time.Hour
```
Comment (liamsi) [inline p2p/client.go:27]: That is so long because once they've advertized this it won't change?

**Comment by liamsi** [comment]
It seems the nodes are not really connected with each other now.

File: p2p/client.go (hunk 5 above)
Comment (tzdybal) [reply p2p/client.go:27]: Exactly.

File: p2p/client_test.go (hunk 4 above)
Comment (tzdybal) [reply p2p/client_test.go:133]: This is a work in progress, big value to give me more time in debugger ;)

Comment (liamsi) [reply p2p/client_test.go:140]: Doesn't a breakpoint stop the "world"? Or does that depend on which debugger you are using?

File: node/node.go [hunk 6]
```diff
@@ -19,15 +22,19 @@ type Node struct {
...
+
+	Mempool mempool.Mempool
 
 	// keep context here only because of API compatibility
 	// - it's used in `OnStart` (defined in service.Service interface)
 	ctx context.Context
 }
 
-func NewNode(ctx context.Context, conf config.NodeConfig, nodeKey crypto.PrivKey, clientCreator proxy.ClientCreator, logger log.Logger) (*Node, error) {
+func NewNode(ctx context.Context, conf config.NodeConfig, nodeKey crypto.PrivKey, clientCreator proxy.ClientCreator, genesis *types.GenesisDoc, logger log.Logger) (*Node, error) {
```
Comment (liamsi) [inline node/node.go:37]: optional nit: IMO a line break per parameter would increase readability.

File: p2p/client.go [hunk 7]
```diff
@@ -178,20 +210,22 @@ func (c *Client) peerDiscovery(ctx context.Context) error {
 
 func (c *Client) setupPeerDiscovery(ctx context.Context) error {
 	// wait for DHT
-	<-c.dht.RefreshRoutingTable()
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.dht.RefreshRoutingTable():
+	}
```
Comment (Wondertan) [inline p2p/client.go:217]: 👍🏻

File: p2p/client.go [hunk 8]
```diff
@@ -13,15 +13,32 @@ import (
...
+	peerLimit = 60
+
+	// txTopicSuffix is added after namespace to create pubsub topic for TX gossiping.
+	txTopicSuffix = "-tx"
+)
+
+// TODO(tzdybal): refactor. This is only a stub.
+type Tx struct {
+	Data []byte
+}
```
Comment (liamsi) [inline p2p/client.go:40]: IMO, we will most likely keep it that way or even make it an alias like in tendermint: https://github.com/tendermint/tendermint/blob/9ecfcc93a6364bbecc9e1d7740b53602eda667eb/types/tx.go#L14-L17

Reason is the abstraction between tendermint (and with it optimint) and the abci app (that understands the Tx). 

It is certainly not the right place for the type definition though but that will be part of that refactoring mentioned in the todo I guess.

File: p2p/client_test.go [hunk 9]
```diff
@@ -93,25 +95,72 @@ func TestDiscovery(t *testing.T) {
...
+	// TODO(tzdybal): is there a better way to wait for readiness?
+	time.Sleep(1 * time.Second)
+
+	// gossip from client 4
+	err := clients[4].GossipTx(ctx, expectedMsg)
+	assert.NoError(err)
+
+	// wait for clients that should receive Tx
+	wg.Wait()
+}
```
Comment (liamsi) [inline p2p/client_test.go:162]: 👍🏼 Nice job on the test.

**Review by liamsi — APPROVED** [review]
LGMT 👍🏼

File: p2p/client_test.go [hunk 10]
```diff
@@ -93,25 +95,72 @@ func TestDiscovery(t *testing.T) {
...
+	assertNotRecv := func(*Tx) {
+		t.Fatal("unexpected Tx received")
+	}
+	clients[1].SetTxHandler(assertNotRecv)
+	clients[2].SetTxHandler(assertNotRecv)
+	clients[4].SetTxHandler(assertNotRecv)
+
+	// this sleep is required for pubsub to "propagate" subscription information
+	// TODO(tzdybal): is there a better way to wait for readiness?
+	time.Sleep(1 * time.Second)
```
Comment (Wondertan) [inline p2p/client_test.go:154]: Nope) Even tests in libp2p use sleep but with much less time) Though I've seen discussions about fixing this several time thought more than a year)

**Review by Wondertan — APPROVED** [review]

File: p2p/client_test.go (hunk 10 above)
Comment (tzdybal) [reply p2p/client_test.go:154]: 1s is empirical result for GitHub actions ;) Locally even 100ms works.

**Pull request merged**